mlir/include/mlir/Dialect/Vector/IR/VectorOps.td

```diff
@@ -478,6 +478,69 @@ def Vector_ShuffleOp :
   let hasCanonicalizer = 1;
 }
 
+def Vector_InterleaveOp :
+  Vector_Op<"interleave", [Pure,
+    AllTypesMatch<["lhs", "rhs"]>,
+    TypesMatchWith<
+    "type of 'result' is double the width of the inputs",
+    "lhs", "result",
+    [{
+      [&]() -> ::mlir::VectorType {
+        auto vectorType = ::llvm::cast<mlir::VectorType>($_self);
+        ::mlir::VectorType::Builder builder(vectorType);
+        if (vectorType.getRank() == 0) {
+          static constexpr int64_t v2xty_shape[] = { 2 };
+          return builder.setShape(v2xty_shape);
+        }
+        auto lastDim = vectorType.getRank() - 1;
+        return builder.setDim(lastDim, vectorType.getDimSize(lastDim) * 2);
+      }()
+    }]>]> {
+  let summary = "constructs a vector by interleaving two input vectors";
+  let description = [{
+    The interleave operation constructs a new vector by interleaving the
+    elements from the trailing (or final) dimension of two input vectors,
+    returning a new vector where the trailing dimension is twice the size.
+
+    Note that for the n-D case this differs from the interleaving possible with
```

Comment:
Thanks for proposing this, 2 questions come to mind:
1. what is the state of the 1-D support for vector.shuffle / shufflevector in MLIR / LLVM for scalable types? Is this generally supported for various permutations or is interleaving a desirable outlier (because e.g. the HW can really do well on this but not on more general permutations)?
2. would this be necessary if https://github.com/llvm/llvm-project/commit/1fe65688d42d1dacca528a871ac8de370043f793 had introduced shuffles to be on the innermost dim? I can't remember the rationale but shuffling the outer dims is more or less a loop in many cases since these dimensions get unrolled when lowering, on most HW...

I am trying to gauge whether we would be better off having a `vector.shuffle_innermost_dim` that takes a general shuffle mask, for which `vector.interleave` is a special case that is easy to detect.

In the absence of compelling HW reasons to only have interleave, I'd favor a "better shuffle" because it is very easy to detect that `[0, n, 1, n+1 ... n-1, 2n-1]` is an interleave.

Reply:
See also the `shufflevector` docs: https://llvm.org/docs/LangRef.html#id189

> For scalable vectors, the only valid mask values at present are zeroinitializer, undef and poison, since we cannot write all indices as literals for a vector with a length unknown at compile time.

Reply:
So, in short:

1. There's basically no support for `shufflevector`, `vector.shuffle`, etc for scalable vectors 
2. I believe so 

For `2`., while you can hack it and claim to arbitrary pattern means interleave for scalable vectors, there's no pattern that'd actually make sense with the current semantics of `vector.shuffle`. So I think that's just adding tricky edge cases in the meaning of the op.

Reply:
Yes, plus this is also similar to having a `vector.broadcast` vs just using a shuffle for that. Interleaving/deinterleaving are very common patterns and having something with a bit higher level is helpful.